stdlib/src/builtin/string.mojo

```diff
@@ -284,10 +284,7 @@ fn _atol(str_ref: StringRef, base: Int = 10) raises -> Int:
 
     var found_valid_chars_after_start = False
     var has_space_after_number = False
-    # single underscores are only allowed between digits
-    # starting "was_last_digit_undescore" to true such that
-    # if the first digit is an undesrcore an error is raised
-    var was_last_digit_undescore = True
+    var was_last_digit_undescore = real_base == 10
```

Comment:
You're right, I hadn't seen `"Base must be >= 2 and <= 36, or 0."` so that means this could receive other bases that have no prefix. So yeah totally, leading underscore should fail
Python:
```text
print(int("_123", 30))
ValueError: invalid literal for int() with base 30: '_123'
```
so it should be 
```mojo
real_base in (2, 8, 16)
```
nice catch, maybe we could add some tests with assert raises (and also passing ones) with bases other than the classics

Reply:
Will for sure add some test cases with the next commit. We will have to add a check for whether there exists a prefix as well though. 

So something along the lines of `real_base in (2, 8, 16) and has_prefix`, as we could have `atol("_ff", 16)` which should raise as invalid. 

Was playing around with `(real_base & 0b11010) != real_base and has_prefix` which would've have been fun... 10 and 18 breaks it though.

Reply:
> So something along the lines of real_base in (2, 8, 16) and has_prefix, as we could have atol("_ff", 16) which should raise as invalid.

yeah I guess `str_ref[start] == "0"` can be taken out of the if to a has_prefix var, nice catch again :D

> Was playing around with (real_base & 0b11010) != real_base and has_prefix which would've have been fun...

Niiice, maybe this will work?
```mojo
from bit import is_power_of_two
...
# first character as underscore is only valid for bases 2, 8, 16 with prefix
var was_last_digit_undescore = is_power_of_two(real_base & 0b11010) and has_prefix:
    ...
```

Reply:
Unfortunately, 10 and 18 breaks this as well. I like the thought though.

We can use a mask,  `bitmask = (1 << 2) | (1 << 8) | (1 << 16)` and then `not ((1 << real_base) & bitmask) and has_prefix`. 

This seems to work. But it might be overcomplicated for a simple value range check.

Reply:
> This seems to work. But it might be overcomplicated for a simple value range check.

yeah I think this was a fun exercise lol but let's just stick with tuple contains

just out of curiosity, does this work ?
```mojo
var mask = (0b11010 * (real_base < 17 and has_prefix).cast[DType.uint8]())
var was_last_digit_undescore = not is_power_of_two(real_base & mask)
```

Reply:
> yeah I think this was a fun exercise lol but let's just stick with tuple contains

I agree. At least we can defer further optimisation to a later PR should we find it necessary.

Reply:
@martinvuyk, I couldn't get it to run. Cannot call `cast` on a Boolean.

Reply:
Then it will probably work by just putting `int(real_base < 17 and has_prefix)`

Reply:
Same issue @martinvuyk, but now for `Int`.

Reply:
yeah but removing the `.cast` since it is a SIMD method